File: Catch Tests/TimeOfDayMathTests.m
#import <XCTest/XCTest.h>
#import "NSDate+TimeOfDayMath.h"


@interface TimeOfDayMathTests : XCTestCase @end

@implementation TimeOfDayMathTests

- (NSDate *)timeWithHour:(NSInteger)hour minute:(NSInteger)minute {
    NSDateComponents *components = NSDateComponents.new;
    
    components.calendar = NSCalendar.currentCalendar;
    
    components.hour = hour;
    components.minute = minute;
    
    return components.date;
}

- (void)testNonWrappingIntervals {
    NSDate *time0415 = [self timeWithHour:4 minute:15];
    NSDate *time0400 = [self timeWithHour:4 minute:0];
    NSDate *time0500 = [self timeWithHour:5 minute:0];
    XCTAssertTrue([time0415 isTimeOfDayBetweenDate:time0400 andDate:time0500]);
    XCTAssertFalse([time0400 isTimeOfDayBetweenDate:time0415 andDate:time0500]);
    XCTAssertFalse([time0500 isTimeOfDayBetweenDate:time0400 andDate:time0415]);
}

- (void)testWrappingIntervals {
    NSDate *time0715 = [self timeWithHour:7 minute:15];
    NSDate *time0245 = [self timeWithHour:2 minute:45];
    NSDate *time0700 = [self timeWithHour:7 minute:0];
    NSDate *time0300 = [self timeWithHour:3 minute:0];
    XCTAssertTrue([time0715 isTimeOfDayBetweenDate:time0700 andDate:time0300]);
    XCTAssertTrue([time0245 isTimeOfDayBetweenDate:time0700 andDate:time0300]);
    XCTAssertFalse([time0700 isTimeOfDayBetweenDate:time0715 andDate:time0300]);
    XCTAssertFalse([time0300 isTimeOfDayBetweenDate:time0700 andDate:time0245]);
}

- (void)testExactTimes {
    // Upper bound is considered "outside" of range
    NSDate *time0400 = [self timeWithHour:4 minute:0];
    NSDate *time0500 = [self timeWithHour:5 minute:0];
    XCTAssertTrue([time0400 isTimeOfDayBetweenDate:time0400 andDate:time0500]);
    XCTAssertFalse([time0500 isTimeOfDayBetweenDate:time0400 andDate:time0500]);
    
    NSDate *time0900 = [self timeWithHour:9 minute:0];
    NSDate *time0100 = [self timeWithHour:1 minute:0];
    XCTAssertTrue([time0900 isTimeOfDayBetweenDate:time0900 andDate:time0100]);
    XCTAssertFalse([time0100 isTimeOfDayBetweenDate:time0900 andDate:time0100]);
}

@end

File: Sources/Tests/TimeOfDayMathTests.swift
import XCTest
@testable import Catch


private extension Date {
  static func timeOfDay(hour: Int, minute: Int) -> Date {
    return DateComponents(calendar: .current, hour: hour, minute: minute).date!
  }
}


class TimeOfDayMathTests: XCTestCase {
  func testNonWrappingIntervals() {
    let time0415 = Date.timeOfDay(hour: 4, minute: 15)
    let time0400 = Date.timeOfDay(hour: 4, minute: 0)
    let time0500 = Date.timeOfDay(hour: 5, minute: 0)
    XCTAssertTrue(time0415.isTimeOfDayBetween(startTimeOfDay: time0400, endTimeOfDay: time0500))
    XCTAssertFalse(time0400.isTimeOfDayBetween(startTimeOfDay: time0415, endTimeOfDay: time0500))
    XCTAssertFalse(time0500.isTimeOfDayBetween(startTimeOfDay: time0400, endTimeOfDay: time0415))
  }
  
  func testWrappingIntervals() {
    let time0715 = Date.timeOfDay(hour: 7, minute: 15)
    let time0245 = Date.timeOfDay(hour: 2, minute: 45)
    let time0700 = Date.timeOfDay(hour: 7, minute: 0)
    let time0300 = Date.timeOfDay(hour: 3, minute: 0)
    XCTAssertTrue(time0715.isTimeOfDayBetween(startTimeOfDay: time0700, endTimeOfDay: time0300))
    XCTAssertTrue(time0245.isTimeOfDayBetween(startTimeOfDay: time0700, endTimeOfDay: time0300))
    XCTAssertFalse(time0700.isTimeOfDayBetween(startTimeOfDay: time0715, endTimeOfDay: time0300))
    XCTAssertFalse(time0300.isTimeOfDayBetween(startTimeOfDay: time0700, endTimeOfDay: time0245))
  }
  
  func testExactTimes() {
    // Upper bound is considered "outside" of range
    let time0400 = Date.timeOfDay(hour: 4, minute: 0)
    let time0500 = Date.timeOfDay(hour: 5, minute: 0)
    XCTAssertTrue(time0400.isTimeOfDayBetween(startTimeOfDay: time0400, endTimeOfDay: time0500))
    XCTAssertFalse(time0500.isTimeOfDayBetween(startTimeOfDay: time0400, endTimeOfDay: time0500))
    
    let time0900 = Date.timeOfDay(hour: 9, minute: 0)
    let time0100 = Date.timeOfDay(hour: 1, minute: 0)
    XCTAssertTrue(time0900.isTimeOfDayBetween(startTimeOfDay: time0900, endTimeOfDay: time0100))
    XCTAssertFalse(time0100.isTimeOfDayBetween(startTimeOfDay: time0900, endTimeOfDay: time0100))
  }
}
